rplugin/python3/deoplete/sources/tabnine.py

```diff
@@ -144,8 +145,15 @@ def _request(self, name, **params):
             self._restart()
             return
 
-        output = proc.stdout.readline()
         try:
+            # try to read from proc's stdout
+            events = selector.select(timeout=0.5)
+            if len(events) == 0:
+                # nothing from TabNine. Restart it
+                self._restart()
```

Comment:
This is rather aggressive, TabNine tends to not return results in when the prediction is for a newline. So this we will cause way more restarts than needed and will damage the experience.

I would also like to point you to our implementation for that in IntelliJ:
https://github.com/codota/tabnine-intellij/blob/b8ddb130ec007a757e6fc181a9bc745cf946bac7/src/main/java/com/tabnine/binary/TabNineGateway.java#L98

Here we validate results based on schema, and only upon a few consecutive bad results we restart the process.

Reply:
As far as I understand, deoplete does not request a completion for an empty line, unless specifically configured to do so.

Reply:
Another question; how do you keep the process of asking for completions and getting replies in sync if you ignore a bad respons?

Reply:
If there is a bas result (exception level bad) we restart the binary. If the response is just not parsing correctly or returns as null we try again (up to 5 consecutive times, on the 5th time we restart).
Please consult the tabnine-intellij plugin and vscode plugin for implementation details.

P. S. Sorry for the late response, my email is bloated AF.

Reply:
The solution as it currently stands is simpler. Do you have any statistics on the average number of tries needed in the intelij plugin? I would bet it's heavily skewed towards working on the first time or failing all 5 attempts. If so, I don't see the need to complicate the code when a simple restart works.

Reply:
Yep your are right, but 0 results is a valid result. `null` is also a valid result in some cases...

Reply:
The problem is we need to distinguish between an empty reply and the daemon dying on us or taking too long to reply.
The current approach works. Why change it?

Reply:
Your call, I explained the constraints. It sure can work, it just do much more restarts than needed, that might impact the experience.